File: mkdocs/snippets/devbook/transactions/transfer_mosaics.mjs
import { PrivateKey } from 'symbol-sdk';
import {
	Address,
	NemFacade,
	calculateTransactionFee,
	descriptors,
	models
} from 'symbol-sdk/nem';

const NODE_URL = process.env.NODE_URL ||
	'http://libertalia.nemtest.net:7890';
console.log('Using node', NODE_URL);

// Helper function to announce a transaction
async function announceTransaction(payload, label) {
	const announcePath = '/transaction/announce';
	console.log(`Announcing ${label} to ${announcePath}`);
	const response = await fetch(`${NODE_URL}${announcePath}`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: payload
	});
	if (!response.ok)
		throw new Error(`HTTP ${response.status}`);
	const result = await response.json();
	console.log('  Result:', result.message);
	return result.message;
}

// Helper function to wait for transaction confirmation
async function waitForConfirmation(transactionHash, label) {
	const statusPath = `/transaction/get?hash=${transactionHash}`;
	console.log(`Waiting for ${label} confirmation from`, statusPath);
	let isConfirmed = false;
	for (let attempt = 1; 120 >= attempt; ++attempt) {
		const response = await fetch(`${NODE_URL}${statusPath}`);
		if (!response.ok) {
			console.log('  Transaction status: pending');
			await new Promise(resolve => { setTimeout(resolve, 1000); });
		} else {
			const confirmed = await response.json();
			console.log(`${label} confirmed in block`,
				confirmed.meta.height);
			isConfirmed = true;
			break;
		}
	}
	if (!isConfirmed)
		console.warn(`${label} confirmation took too long.`);
}

// [>step-1]
const SIGNER_PRIVATE_KEY = process.env.SIGNER_PRIVATE_KEY ||
	'0000000000000000000000000000000000000000000000000000000000000000';
const signerKeyPair = new NemFacade.KeyPair(
	new PrivateKey(SIGNER_PRIVATE_KEY));

const RECIPIENT_ADDRESS = process.env.RECIPIENT_ADDRESS ||
	'TBULEAUG2CZQISUR442HWA6UAKGWIXHDABJVIPS4';
// [<step-1]
// [>step-2]
const MOSAIC_ID = process.env.MOSAIC_ID || 'company:token';
const [MOSAIC_NAMESPACE, MOSAIC_NAME] = MOSAIC_ID.split(':');
const QUANTITY = parseInt(process.env.QUANTITY || '100', 10);
console.log('Sending mosaic', MOSAIC_ID);
console.log(`  Amount: ${QUANTITY} units`);
// [<step-2]
const facade = new NemFacade('testnet');

try {
	// Fetch the mosaic's divisibility and supply [>step-3]
	const definitionPath = `/mosaic/definition?mosaicId=${MOSAIC_ID}`;
	console.log('Fetching mosaic definition from', definitionPath);
	const definitionResponse =
		await fetch(`${NODE_URL}${definitionPath}`);
	const definition = await definitionResponse.json();
	const properties = Object.fromEntries(
		definition.properties.map(
			property => [property.name, property.value]));
	const divisibility = parseInt(properties.divisibility, 10);

	const supplyPath = `/mosaic/supply?mosaicId=${MOSAIC_ID}`;
	console.log('Fetching mosaic supply from', supplyPath);
	const supplyResponse = await fetch(`${NODE_URL}${supplyPath}`);
	const { supply } = await supplyResponse.json();
	console.log(`  ${MOSAIC_ID}: divisibility ${divisibility},`,
		`supply ${supply}`);
	// [<step-3]
	// Build the transaction [>step-4]
	const atomicQuantity = QUANTITY * (10 ** divisibility);
	const multiplier = 1;
	const scaledMultiplier = multiplier * 1_000_000;
	const transaction = facade.createTransactionFromTypedDescriptor(
		new descriptors.TransferTransactionV2Descriptor(
			new Address(RECIPIENT_ADDRESS),
			new models.Amount(BigInt(scaledMultiplier)),
			undefined,
			[new descriptors.SizePrefixedMosaicDescriptor(
				new descriptors.MosaicDescriptor(
					new descriptors.MosaicIdDescriptor(
						new descriptors.NamespaceIdDescriptor(
							MOSAIC_NAMESPACE),
						MOSAIC_NAME),
					new models.Amount(BigInt(atomicQuantity))))]),
		signerKeyPair.publicKey,
		0n,
		2 * 60 * 60);
	// [<step-4]
	// Calculate and attach the transaction fee [>step-5]
	const fee = calculateTransactionFee(transaction, {
		[MOSAIC_ID]: { supply: BigInt(supply), divisibility }
	});
	transaction.fee = new models.Amount(fee);
	console.log(`  Transaction fee: ${Number(fee) / 1_000_000} XEM`);
	// [<step-5]
	// Sign transaction and generate final payload [>step-6]
	const signature = facade.signTransaction(signerKeyPair, transaction);
	const jsonPayload = facade.transactionFactory.static.attachSignature(
		transaction, signature);
	console.log('Built transaction:');
	console.dir(transaction.toJson(), { colors: true, depth: null });
	// [<step-6]
	// Announce the transaction [>step-7]
	const announceResult = await announceTransaction(
		jsonPayload, 'transaction');
	// [<step-7]
	// Wait for confirmation [>step-8]
	if ('SUCCESS' === announceResult) {
		const transactionHash = facade.hashTransaction(transaction)
			.toString();
		console.log('Transaction hash:', transactionHash);
		await waitForConfirmation(transactionHash, 'transaction');
	} else {
		console.log('Transaction rejected:', announceResult);
	}
	// [<step-8]
} catch (e) {
	console.error(e.message, '| Cause:', e.cause?.code ?? 'unknown');
}

File: mkdocs/snippets/devbook/transactions/TransferMosaics.java
//JAVA 21+
//DEPS org.symbol:symbol-sdk:3.3.3

import static java.nio.charset.StandardCharsets.UTF_8;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.symbol.sdk.CryptoTypes;
import org.symbol.sdk.facade.NemFacade;
import org.symbol.sdk.nem.Address;
import org.symbol.sdk.nem.FeeCalculator;
import org.symbol.sdk.nem.FeeCalculator.MosaicInformation;
import org.symbol.sdk.nem.KeyPair;
import org.symbol.sdk.nem.NemTransactionFactory;
import org.symbol.sdk.nem.descriptors.MosaicDescriptor;
import org.symbol.sdk.nem.descriptors.MosaicIdDescriptor;
import org.symbol.sdk.nem.descriptors.NamespaceIdDescriptor;
import org.symbol.sdk.nem.descriptors.SizePrefixedMosaicDescriptor;
import org.symbol.sdk.nem.descriptors.TransferTransactionV2Descriptor;
import org.symbol.sdk.nem.models.Amount;
import org.symbol.sdk.nem.models.Transaction;

public final class TransferMosaics {
	private static final ObjectMapper JSON_MAPPER = new ObjectMapper();

	private static final HttpClient HTTP_CLIENT =
		HttpClient.newHttpClient();

	private final String nodeUrl = System.getenv().getOrDefault(
		"NODE_URL", "http://libertalia.nemtest.net:7890");

	private final NemFacade facade = new NemFacade("testnet");

	private String announceTransaction(
		final String payload,
		final String label
	) throws IOException, InterruptedException {
		final String announcePath = "/transaction/announce";
		System.out.printf("Announcing %s to %s%n", label, announcePath);
		final HttpRequest request = HttpRequest.newBuilder(
			URI.create(nodeUrl + announcePath))
			.header("Content-Type", "application/json")
			.POST(HttpRequest.BodyPublishers.ofString(payload))
			.build();
		final HttpResponse<String> response = HTTP_CLIENT.send(
			request, BodyHandlers.ofString());
		if (2 != response.statusCode() / 100)
			throw new IOException("HTTP " + response.statusCode());
		final String result = JSON_MAPPER.readTree(response.body())
			.get("message").asText();
		System.out.printf("  Result: %s%n", result);
		return result;
	}

	private void waitForConfirmation(
		final String transactionHash,
		final String label
	) throws IOException, InterruptedException {
		final String statusPath = String.format(
			"/transaction/get?hash=%s", transactionHash);
		System.out.printf("Waiting for %s confirmation from %s%n",
			label, statusPath);
		boolean isConfirmed = false;
		for (int attempt = 1; 120 >= attempt; ++attempt) {
			final HttpRequest request = HttpRequest.newBuilder(
				URI.create(nodeUrl + statusPath)).GET().build();
			final HttpResponse<String> response = HTTP_CLIENT.send(
				request, BodyHandlers.ofString());
			if (2 != response.statusCode() / 100) {
				System.out.println("  Transaction status: pending");
				Thread.sleep(1000);
			} else {
				final JsonNode confirmed = JSON_MAPPER.readTree(
					response.body());
				System.out.printf("%s confirmed in block %s%n", label,
					confirmed.get("meta").get("height").asText());
				isConfirmed = true;
				break;
			}
		}
		if (!isConfirmed)
			System.out.printf("%s confirmation took too long.%n", label);
	}

	public static void main(final String[] args) {
		try {
			new TransferMosaics().run();
		} catch (final Exception ex) {
			System.out.println(null == ex.getMessage()
				? ex.toString()
				: ex.getMessage());
		}
	}

	private void run() throws IOException, InterruptedException {
		System.out.printf("Using node %s%n", nodeUrl);

		// [>step-1]
		final String signerPrivateKey = System.getenv().getOrDefault(
			"SIGNER_PRIVATE_KEY", "0".repeat(64));
		final KeyPair signerKeyPair = new KeyPair(
			new CryptoTypes.PrivateKey(signerPrivateKey));
		final String recipientAddress = System.getenv().getOrDefault(
			"RECIPIENT_ADDRESS",
			"TBULEAUG2CZQISUR442HWA6UAKGWIXHDABJVIPS4");
		// [<step-1]
		// [>step-2]
		final String mosaicId = System.getenv().getOrDefault(
			"MOSAIC_ID", "company:token");
		final String[] mosaicIdParts = mosaicId.split(":");
		final String mosaicNamespace = mosaicIdParts[0];
		final String mosaicName = mosaicIdParts[1];
		final int quantity = Integer.parseInt(
			System.getenv().getOrDefault("QUANTITY", "100"));
		System.out.printf("Sending mosaic %s%n", mosaicId);
		System.out.printf("  Amount: %d units%n", quantity);
		// [<step-2]

		// Fetch the mosaic's divisibility and supply [>step-3]
		final String definitionPath = String.format(
			"/mosaic/definition?mosaicId=%s", mosaicId);
		System.out.printf("Fetching mosaic definition from %s%n",
			definitionPath);
		final HttpRequest definitionRequest = HttpRequest.newBuilder(
			URI.create(nodeUrl + definitionPath)).GET().build();
		final HttpResponse<String> definitionResponse = HTTP_CLIENT.send(
			definitionRequest, BodyHandlers.ofString());
		final JsonNode definition = JSON_MAPPER.readTree(
			definitionResponse.body());
		int divisibility = 0;
		for (final JsonNode property : definition.get("properties")) {
			if ("divisibility".equals(property.get("name").asText()))
				divisibility = property.get("value").asInt();
		}
		final String supplyPath = String.format(
			"/mosaic/supply?mosaicId=%s", mosaicId);
		System.out.printf("Fetching mosaic supply from %s%n", supplyPath);
		final HttpRequest supplyRequest = HttpRequest.newBuilder(
			URI.create(nodeUrl + supplyPath)).GET().build();
		final HttpResponse<String> supplyResponse = HTTP_CLIENT.send(
			supplyRequest, BodyHandlers.ofString());
		final long supply = JSON_MAPPER.readTree(supplyResponse.body())
			.get("supply").asLong();
		System.out.printf("  %s: divisibility %d, supply %d%n",
			mosaicId, divisibility, supply);
		// [<step-3]
		// Build the transaction [>step-4]
		final long atomicQuantity = Math.round(
			quantity * Math.pow(10, divisibility));
		final long scaledMultiplier = 1_000_000;
		final MosaicIdDescriptor mosaicIdDescriptor =
			new MosaicIdDescriptor(
				new NamespaceIdDescriptor(mosaicNamespace.getBytes(UTF_8)),
				mosaicName.getBytes(UTF_8));
		final Transaction transaction =
			facade.createTransactionFromTypedDescriptor(
				new TransferTransactionV2Descriptor(
					new Address(recipientAddress),
					new Amount(scaledMultiplier),
					null,
					List.of(new SizePrefixedMosaicDescriptor(
						new MosaicDescriptor(
							mosaicIdDescriptor,
							new Amount(atomicQuantity))))),
				signerKeyPair.getPublicKey(),
				0,
				2 * 60 * 60);
		// [<step-4]
		// Calculate and attach the transaction fee [>step-5]
		final long fee = FeeCalculator.calculateTransactionFee(
			transaction, Map.of(mosaicId,
				new MosaicInformation(supply, divisibility)));
		transaction.setFee(new Amount(fee));
		System.out.printf("  Transaction fee: %s XEM%n",
			(double) fee / 1_000_000);
		// [<step-5]
		// Sign transaction and generate final payload [>step-6]
		final CryptoTypes.Signature signature = facade.signTransaction(
			signerKeyPair, transaction);
		final String jsonPayload = NemTransactionFactory.attachSignature(
			transaction, signature);
		System.out.println("Built transaction:");
		System.out.println(JSON_MAPPER.writerWithDefaultPrettyPrinter()
			.writeValueAsString(transaction.toJson()));
		// [<step-6]
		// Announce the transaction [>step-7]
		final String announceResult = announceTransaction(
			jsonPayload, "transaction");
		// [<step-7]
		// Wait for confirmation [>step-8]
		if ("SUCCESS".equals(announceResult)) {
			final String transactionHash = facade.hashTransaction(
				transaction).toString();
			System.out.printf("Transaction hash: %s%n", transactionHash);
			waitForConfirmation(transactionHash, "transaction");
		} else {
			System.out.printf(
				"Transaction rejected: %s%n", announceResult);
		}
		// [<step-8]
	}
}
